storey/sources.py

```diff
@@ -622,49 +621,78 @@ def __init__(
         WithUUID.__init__(self)
         if isinstance(dfs, pandas.DataFrame):
             dfs = [dfs]
+        if dfs:
+            for df in dfs:
+                self._validate_fields(df=df, key_field=key_field, id_field=id_field)
         self._dfs = dfs
         self._key_field = key_field
         self._id_field = id_field
 
+    def _get_key(self, body):
```

Comment:
The contract of this method is rather unintuitive. Needs to be refactored or documented.